test/support/mock_server.js

```diff
@@ -0,0 +1,149 @@
+
+ScriptedServer.prototype.start = function() {
+    var that = this;
+    this.stream = net.createServer(function(client) {
+        that.emit('connect_request');
+        sendInfo(that, client);
+        client.script = Array.from(that.script);
+        client.on('data', handleData(that, client));
+    });
+
+    this.stream.on('connection', (socket) => {
+        this.sockets.push(socket);
+    });
+
+    this.stream.on('close', () => {
+        this.emit('close');
+    });
+
+    this.stream.on('error', (ex) => {
+        this.emit('error', ex);
+    });
+
+    this.stream.on('listening', () => {
+        this.emit('listening');
+    });
+
+    this.stream.listen(that.port);
+};
+
+ScriptedServer.prototype.stop = function(cb) {
+    this.stream.close(cb);
+    this.sockets.forEach(function(socket) {
+        if (!socket.destroyed) {
+            socket.destroy();
+        }
+    });
+};
+
+function handleData(server, client) {
+    return function(data) {
+        // if we have a buffer append to it or make one
+        if (client.buffer) {
+            client.buffer = Buffer.concat([client.buffer, data]);
+        } else {
+            client.buffer = data;
+        }
+
+        // convert to string like node-nats does so we can test protocol
+        var buf = client.buffer.toString('binary', 0, MAX_CONTROL);
+        if (client.script.length) {
+            var match;
+            if (match = client.script[0].re.exec(buf)) {
+                // if we have a match, execute the handler
+                client.script[0].h(client, match);
+
+                // prune the buffer without the processed request
+                var len = match[0].length;
+                if (len >= client.buffer.length) {
+                    delete client.buffer;
+                } else {
+                    client.buffer = client.buffer.slice(len);
+                }
+
+                // delete the script step
+                client.script.shift();
+            } else {
+                server.emit('warn', 'no match:\n' + colorize(buf))
+            }
+        } else {
+            server.emit('info','no more script handlers, ignoring:\n' + colorize(buf))
+        }
+    }
+}
+
+function colorize(str) {
+    return str.replace(/(?:\r\n)/g, '\\r\\n\n');
+}
+
+function sendInfo(server, socket) {
+    socket.write("INFO " + JSON.stringify({
+        server_id: "TEST",
+        version: "0.0.0",
+        node: "node0.0.0",
+        host: "127.0.0.1",
+        port: server.port,
+        auth_required: false,
+        ssl_required: false,
+        tls_required: false,
+        tls_verify: false,
+        max_payload: MAX_CONTROL
+    }) + "\r\n");
+}
+
+function sendOk(socket) {
+    socket.write("+OK\r\n");
+}
+
+exports.sendOk = sendOk;
+
+function sendPing(socket) {
+    socket.write("PING\r\n");
+}
+
+exports.sendPing = sendPing;
+
+function sendPong(socket) {
+    socket.write("PONG\r\n");
+}
+
+exports.sendPong = sendPong;
```

Comment:
nit: do you need new line here?

Reply:
I added it as I showed you. For some reason, they are getting stripped out of the diff.